content/6_Advanced/String_Search.mdx

```diff
@@ -414,6 +407,129 @@ Build the entire trie first, and then run a _BFS_ to construct the fail array.
 	 />
 </Resources>
 
+The Aho-Corasick algorithm stores the pattern words in a trie structure, described above.
+It uses the trie to transition from a state to anoother.
+Similar to KMP algorithm, we want to reuse the information we have already processed.
+
+A **suffix link** or **failure link** for a node $u$ is a special edge that points to the longest proper suffix of the string corresponding to node $u$.
+The suffix links for the root and all its immediate children point to the root node.
+For all other nodes $u$ with parent $p$ and letter $c$ on the edge $p \rightarrow u$ the suffix link
+can be computed by following $p$'s failure link and transitioning to letter $c$ from there.
+
+While processing the string $S$ the algorithm maintains the current node in the trie such that word
+formed in the node is equal to the longest suffix ending in $i$.
+
+For example, when transitioning from $i$ to $i+1$ in $S$ there only are two choices:
+1. If $node$ does have an outgoing edge with letter $S_{i+1}$, then move down the edge.
+2. Otherwise, follow the failure link of $S_i$ and transition to letter $S_{i+1}$ from there.
+
+The image below depicts how the structure looks like for the words $[a, ag, c, caa, gag, gc]$.
+
+<center>
+![Trie](./Trie.png)
+<i> An Aho-Corasick trie with failure links as light edges. </i>
+</center>
+
+There is a special case when some words are substring of other words in the wordlist. This could lead to some problems depending on the implementation. [Dictionary links](https://www.youtube.com/watch?v=OFKxWFew_L0) can solve this problem.
+They act like suffix links that point to the first suffix that is also a word in the wordlist. The code below constructs the structure using a BFS.
+
+**Time Complexity:** $\mathcal{O}(m\sigma)$ - where $m$ is the size of the alphaber and $\sigma$ the size of the alphabet
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+
+using namespace std;
+
+const int NMAX = 6e5;
+const int SIGMA = 26;
+
+int n;
+string s;
+int nodes = 1;
+int trie[NMAX][SIGMA];
+int fail[NMAX], seen[NMAX], ans[NMAX];
```

Comment:
comments for `leaf` but not any of these?

Reply:
Done